From 612473584528920a34eed38336bfbbd6c0955598 Mon Sep 17 00:00:00 2001
From: Ye Bin <yebin10@huawei.com>
Date: Sun, 28 Mar 2021 23:13:06 -0400
Subject: [PATCH] arm32: kaslr: Fix clock_gettime and gettimeofday performance
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 612473584528920a34eed38336bfbbd6c0955598

 degradation when configure CONFIG_RANDOMIZE_BASE

hulk inclusion
category: feature
feature: ARM kaslr support
bugzilla: 47952
CVE: NA

-----------------------------------------------

When we configure CONFIG_RANDOMIZE_BASE we find that:
[XX]$arm-linux-gnueabihf-readelf -s  ./arch/arm/vdso/vdso.so
Symbol table '.dynsym' contains 5 entries:
   Num:    Value  Size Type    Bind   Vis      Ndx Name
	 0: 00000000     0 NOTYPE  LOCAL  DEFAULT  UND
	 1: 00000278     0 SECTION LOCAL  DEFAULT    8
	 2: 00000000     0 OBJECT  GLOBAL DEFAULT  ABS LINUX_2.6

  We can't find __vdso_gettimeofday and __vdso_clock_gettime symbol. So call
clock_gettime and gettimeofday will call system call. This results in
performance degradation.

Signed-off-by: Ye Bin <yebin10@huawei.com>
Reviewed-by: Jason Yan <yanaijie@huawei.com>
Signed-off-by: yangerkun <yangerkun@huawei.com>
Signed-off-by: Cui GaoSheng <cuigaosheng1@huawei.com>
Reviewed-by: Xiu Jianfeng <xiujianfeng@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm/vdso/vgettimeofday.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/arm/vdso/vgettimeofday.c b/arch/arm/vdso/vgettimeofday.c
index 1976c6f325a4..425e4f2458ec 100644
--- a/arch/arm/vdso/vgettimeofday.c
+++ b/arch/arm/vdso/vgettimeofday.c
@@ -4,6 +4,11 @@
  *
  * Copyright 2015 Mentor Graphics Corporation.
  */
+
+#ifdef CONFIG_RANDOMIZE_BASE
+#pragma GCC visibility pop
+#endif
+
 #include <linux/time.h>
 #include <linux/types.h>
 
-- 
2.26.2

